euphonic/spectra.py

```diff
@@ -669,10 +674,387 @@ def broaden(self: T, x_width,
+                 if key not in combined_metadata}
+            for metadata in all_metadata
+        ]
+        if any(line_data):
+            combined_metadata['line_data'] = line_data
+
+        return combined_metadata
+
+    def _tidy_metadata(self) -> Metadata:
+        """
+        For a metadata dictionary, combines all common key/value
+        pairs in 'line_data' and puts them in a top-level dictionary.
+        """
+        line_data = self.metadata.get("line_data", [{}] * len(self))
+        combined_line_data = self._combine_metadata(line_data)
+        combined_line_data.pop("line_data", None)
+        return combined_line_data
+
+    def group_by(self, *line_data_keys: str) -> Self:
+        """
+        Group and sum elements of spectral data according to the values
+        mapped to the specified keys in metadata['line_data']
+
+        Parameters
+        ----------
+        line_data_keys
+            The key(s) to group by. If only one line_data_key is
+            supplied, if the value mapped to a key is the same for
+            multiple spectra, they are placed in the same group and
+            summed. If multiple line_data_keys are supplied, the values
+            must be the same for all specified keys for them to be
+            placed in the same group
+
+        Returns
+        -------
+        grouped_spectrum
+            A new Spectrum1DCollection with one line for each group. Any
+            metadata in 'line_data' not common across all spectra in a
+            group will be discarded
+        """
+        def get_key_items(enumerated_metadata: tuple[int, OneLineData]
+                          ) -> tuple[str | int, ...]:
+            """Get sort keys from an item of enumerated input to groupby
+
+            e.g. with line_data_keys=("a", "b")
+
+              (0, {"a": 4, "d": 5}) --> (4, None)
+            """
+            return tuple(enumerated_metadata[1].get(item, None)
+                         for item in line_data_keys)
+
+        # First element of each tuple is the index
+        indices = partial(pluck, 0)
+
+        groups = groupby(get_key_items, enumerate(self.iter_metadata()))
+
+        return self.from_spectra([self[list(indices(group))].sum()
+                                  for group in groups.values()])
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Convert to a dictionary consistent with from_dict()
+
+        Returns
+        -------
+        dict
+        """
+        attrs = [*self._get_bin_kwargs().keys(),
+                 self._spectrum_data_name(),
+                 'x_tick_labels',
+                 'metadata']
+
+        return _obj_to_dict(self, attrs)
+
+    @classmethod
+    def from_dict(cls: Self, d: dict) -> Self:
+        """Initialise a Spectrum Collection object from dict"""
+        data_keys = list(f"{dim}_data" for dim in cls._bin_axes)
+        data_keys.append(cls._spectrum_data_name())
+
+        d = _process_dict(d,
```

Comment:
I'm struggling for a good variable name here! Pandas uses `data` as the from_dict argument, but here that's very confusing because "data" is used in the other variable names to mean x_data, y_data, z_data...

Reply:
(The other problem is that I used `d` for consistency with existing methods. If we rename all of them that's a break in public API. It should be handled with backward compatibility and deprecation warning.)

Reply:
Well, best not to break it if there's no real improvement.